src/java/test/org/apache/zookeeper/server/MultiOpSessionUpgradeTest.java

```diff
@@ -0,0 +1,135 @@
+        Assert.assertNotNull("unable to find server interlocutor", server);
+        UpgradeableSessionTracker sessionTracker = (UpgradeableSessionTracker)server.getSessionTracker();
+        Assert.assertFalse("session already global", sessionTracker.isGlobalSession(zk.getSessionId()));
+
+        List<OpResult> multi = null;
+        try {
+            multi = zk.multi(Arrays.asList(
+                    Op.setData(path, data.getBytes(), 0),
+                    Op.create(path + "/e", data.getBytes(), ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.EPHEMERAL),
+                    Op.create(path + "/p", data.getBytes(), ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT),
+                    Op.create(path + "/q", data.getBytes(), ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.EPHEMERAL)
+            ));
+        } catch (KeeperException.SessionExpiredException e) {
+            // the scenario that inspired this unit test
+            Assert.fail("received session expired for a session promotion in a multi-op");
+        }
+
+        Assert.assertNotNull(multi);
+        Assert.assertEquals(4, multi.size());
+        Assert.assertEquals(data, new String(zk.getData(path + "/e", false, null)));
+        Assert.assertEquals(data, new String(zk.getData(path + "/p", false, null)));
+        Assert.assertEquals(data, new String(zk.getData(path + "/q", false, null)));
+        Assert.assertTrue("session not promoted", sessionTracker.isGlobalSession(zk.getSessionId()));
+    }
+
+    @Test
+    public void directCheckUpgradeSessionTest() throws IOException, InterruptedException, KeeperException {
+        final ZooKeeper zk = createClient();
+
+        String path = "/directcheckupgradesession";
+        zk.create(path, new byte[0], ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+        QuorumZooKeeperServer server = getConnectedServer(zk.getSessionId());
+        Assert.assertNotNull("unable to find server interlocutor", server);
+
+        Request readRequest = makeGetDataRequest(path, zk.getSessionId());
+        Request createRequest = makeCreateRequest(path + "/e", zk.getSessionId());
+        Assert.assertNull("tried to upgrade on a read", server.checkUpgradeSession(readRequest));
+        Assert.assertNotNull("failed to upgrade on a create", server.checkUpgradeSession(createRequest));
+        Assert.assertNull("tried to upgrade after successful promotion",
+                server.checkUpgradeSession(createRequest));
+    }
+
+    private Request makeGetDataRequest(String path, long sessionId) throws IOException {
+        ByteArrayOutputStream boas = new ByteArrayOutputStream();
+        BinaryOutputArchive boa = BinaryOutputArchive.getArchive(boas);
+        GetDataRequest getDataRequest = new GetDataRequest(path, false);
+        getDataRequest.serialize(boa, "request");
+        ByteBuffer bb = ByteBuffer.wrap(boas.toByteArray());
+        return new Request(null, sessionId, 1, ZooDefs.OpCode.getData, bb, new ArrayList<>());
```

Comment:
`new ArrayList<>()` would not work in JDK 7 (the compiler can't infer the type w/o explicitly specify the type parameter here). I think you are using JDK 8 so you don't see this. Suggest change this to 'new ArrayList<Id>()' so it works with both JDK 7 and 8. Pre-commit job also failed because of this:  https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/462/console